Fixed GCC10 + F2008 compliance

nwu63 opened the pull request:
## Purpose
I added the `-std=f2003` flag and fixed all the errors that came up. Since I'm on GCC 10, I also fixed the argument mismatch issue.

Note that some warnings still remain:
```
parameterizations.f90:38:61:

                       (X(3, K, J, I) - X(3, K, J, I-1)) ** 2)
                                                             1
Warning: Obsolescent feature: Statement function at (1)
```

As part of the process, I removed the `norm` function definition and replaced it by the F2008 intrinsic `NORM2`. I also replaced one scalar call to `norm` with the `ABS` function which is equivalent for scalars.

## Type of change
What types of change is it?
_Select the appropriate type(s) that describe this PR_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [x] Maintenance update
- [ ] Other (please describe)

## Testing
I updated the config file so the `-std=f95` code will be tested, but since we are on GCC 9 I guess we should just observe that the argument mismatch warnings should no longer show up in the compile log.

## Checklist
_Put an `x` in the boxes that apply._

- [ ] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

nwu63 commented on src/projections.F90:
```diff
@@ -56,7 +56,7 @@ subroutine point_curve(x0, t, k, coef, nctl, ndim, Niter, eps, s, Diff)
   ! Working
   real(kind=realType)   :: val(ndim), deriv(ndim), deriv2(ndim), step, c, dist, p_diff
-  real(kind=realType)   :: grad, hessian,  update, R(ndim), nDist, fval, nfval, pgrad, newPt
+  real(kind=realType)   :: grad, temp(1), hessian,  update, R(ndim), nDist, fval, nfval, pgrad, newPt
```
Is this an okay solution? I create a temporary singleton array and use this to pass value to the subroutine, as recommended [here](https://gcc.gnu.org/gcc-10/porting_to.html).

bernardopacini replied:
This seems okay to me.

eirikurj replied:
In general I am not a fan, but this is the simplest. Would need to think for a better solution.

nwu63 replied:
I removed this and replaced by the `norm` invocation with `ABS` which is equivalent for scalars.

nwu63 commented on src/projections.F90:
```diff
@@ -1496,7 +1498,7 @@ subroutine point_plane(pt, p0, v1, v2, n, sol, n_sol, best_sol)
   ! Now post-process to get the closest one
   best_sol = 1
-  D0 = norm(p0(:, ind(1)) + sol(2, ind(1))*v1(:, ind(1)) + sol(3, ind(1))*v2(:, ind(1)))
+  D0 = norm(p0(:, ind(1)) + sol(2, ind(1))*v1(:, ind(1)) + sol(3, ind(1))*v2(:, ind(1)), 3)
```
I'm assuming the missing parameter here is 3, please double check.

bernardopacini replied:
I am having trouble figuring out where this `norm()` function is. Do you know where it references?

eirikurj replied:
The norm function is defined in `compute_curve.f90`. With 2008 we could replace this with `norm2`. Maybe in a separate PR.

nwu63 replied:
Right, I would prefer a separate PR if we want to update to modern intrinsics.

nwu63 replied:
Never mind I just did it here

bernardopacini commented on config/defaults/config.LINUX_GFORTRAN.mk:
```diff
@@ -5,7 +5,7 @@ RM       = /bin/rm -rf
 # Fortran compiler and flags
 FF90        = gfortran
-FF90_FLAGS  = -fdefault-real-8 -O2 -fPIC
+FF90_FLAGS  = -fdefault-real-8 -O2 -fPIC -std=f95
```
This is likely not within the scope of this PR, but I have read that it is bad practice to use the `-fdefault-real-8` flag. It may not matter for us, but I want to note it.

nwu63 replied:
@eirikurj do you know anything about this? Why is it bad practice? We use it pretty much everywhere I think

eirikurj replied:
Maybe @bernardopacini can elaborate on this, but I guess this can be considered bad practice by some, since it may not always give the same precision. However, while not perfect, in most of our architectures it should not be an issue and I would argue that this currently safeguards us from potential precision issues. For the current implementation, I would not want to remove it until we have addressed that.

bernardopacini replied:
I do not have anything formal to refer to about why it is bad practice, I had simply read that in a few places. I agree that for us it is likely far more problematic to take this out than to leave it in.

bernardopacini reviewed (commented):
Besides the three comments, the rest looks good to me.

eirikurj commented on src/adtProjections.F90:
```diff
@@ -1287,7 +1287,7 @@ subroutine searchQuads(pts, conn, searchPts, nPts, nConn, nSearchPts, faceID, uv
        coor(1:3) = searchPts(:, i)
        coor(4) = 1e20
-       call minDistanceTreeSearchSinglePoint(ADT, coor, intInfo, uvw, dummy, &
+       call minDistTreeSearchSinglePoint(ADT, coor, intInfo, uvw, dummy, &
```
I almost feel that this is a step backwards. Since pyspline does not require tapenade I would be more in favor of using 2003/2008.

nwu63 replied:
Let me update to F2008 then, and maybe update the intrinsics at the same time.

nwu63 commented:
Okay I updated the following
- switched to F2003
- reverted the subroutine name shortening
- switched to using NORM2 intrinsic
- replaced scalar invocation of norm with `ABS`

There are other things that can be replaced by intrinsics, such as the `dotproduct` subroutine, but maybe we should save that for later. A lot of our codes can be modernized though.

bernardopacini commented:
These new updates look good to me, but I want to wait for @eirikurj's approval before merging.
